crates/matrix-sdk-ui/src/sync_service.rs

```diff
@@ -90,6 +111,58 @@ impl SyncTaskSupervisor {
         Self { task, termination_sender }
     }
 
+    async fn offline_check(
+        client: &Client,
+        receiver: &mut Receiver<TerminationReport>,
+    ) -> Option<TerminationReport> {
+        info!("Entering the offline mode");
+
+        let wait_for_termination_report = async {
+            loop {
+                // Since we didn't empty the channel when entering the offline mode in fear that
+                // we might miss a report with the
+                // `TerminationOrigin::Supervisor` origin and the channel might contain stale
+                // reports from one of the sync services, in case both of them have sent a
+                // report, let's ignore all reports we receive from the sync
+                // services.
+                let report =
+                    receiver.recv().await.unwrap_or_else(TerminationReport::supervisor_error);
+
+                match report.origin {
+                    TerminationOrigin::EncryptionSync | TerminationOrigin::RoomList => (),
+                    // Since the sync service aren't running anymore, we can only receive a report
+                    // from the supervisor. It would have probably made sense to have separate
+                    // channels for reports the sync services send and the user can send using the
+                    // `SyncService::stop()` method.
+                    TerminationOrigin::Supervisor => break report,
+                }
+            }
+        };
+
+        let wait_to_be_online = async move {
+            loop {
```

Comment:
Is it OK to do this repeatedly with no pauses or backoff?

Reply:
Yes, the backoff is handled by the network requesting layer.

Reply:
Can we explain that in a comment please?

Reply:
I guess we can add some timeout in case the user configures the `RequestConfig` very aggressively.

I can add a comment, yes. It is our usual `RequestConfig` which uses exponential backoff: https://matrix-org.github.io/matrix-rust-sdk/matrix_sdk/config/struct.RequestConfig.html.

Reply:
Added a timeout, just in case, and a comment in https://github.com/matrix-org/matrix-rust-sdk/pull/4592/commits/7ebdf2ebfb7044f44a9f3291f50ee4e3ceba9240.

Reply:
Thanks <3!